### DISCUSSION TITLE
tailwind 동적 스타일링

### DISCUSSION THREAD

**Discussion by elbica** [YAPP-Github/21st-Web-Team-1-FE#24] [General]
방법 1. 객체에 스타일 다 정의해놓고 key로 접근
단점 : 자동완성 안됨 -> vscode tailwind class Regex로 해결 가능

<img width="406" alt="image" src="https://user-images.githubusercontent.com/74011724/206512840-4ae533bd-c1f4-45cc-86ee-011917065b4b.png">
￼￼￼https://ohdoylerules.com/snippets/autocomplete-tailwind-classes/

웹스톰은 안되는듯..ㅜㅜ (내년에 패치예정이라고 함)

방법 2. classname 내에서 분기처리
단점 : 가독성 떨어짐

방법 3. twin macro 사용 (사용성은 가장 좋음)
단점 : css-in-js 방식이 됨 (좀 더 알아봐야 함)

방법 4. 커스텀 클래스 정의
컴포넌트마다 커스텀 클래스를 따로 만들어야 하는데.. 이방식이 맞나?

**Comment by ojj1123** [comment]
스타일을 어떻게 관리해야 할지 고민이어서 얼른 이야기를 하고 싶은데 일단 지금까지 생각해봤던걸 적어볼게~

### 방법 1. 객체에 스타일 다 정의해놓고 key로 접근
```jsx
const style = {
  large: "rounded-[1rem] py-11 px-15 text-semi-bold",
  medium: "rounded-[0.5rem] py-8 px-13 text-tag",
  default: "",
};
```
우리가 지금 이렇게 사용하고 있는데,
이런식으로 `style` 객체에서 가져다 사용하는게 커스텀 클래스 만들어서 하는거랑 다른게 없어보여서 말이야.
공식문서 보니까 커스텀 클래스 만드는걸 함부로 남용하지 말라고 나오긴 하거든?
> don’t use `@apply` just to make things look “cleaner”
> https://tailwindcss.com/docs/reusing-styles#avoiding-premature-abstraction

우리가 `style` 객체에서 `key`로 접근해서 스타일 가져오는 이유가 중복을 피하려는게 주된 이유라는 생각이 들어!
그래서 일단 이 방법은 일단 보류 해두자.


### 방법 2. classname 내에서 분기처리
classname 내에서 분기처리 한다는게 이런 방식을 말한걸까?

```jsx
<li
  className={`${
    active ? 'bg-blue-500 text-white' : 'bg-white text-black'
  }`}
>
  {selected && <CheckIcon />}
  {person.name}
</li>
```


### 방법 3. twin macro 사용
동적 스타일링이랑 className 우선순위 때문에 twin macro 이걸 고려하고 있는거 같아! twin macro 사용하는 것도 좋은 선택지야!

근데 그 전에 어떻게 tailwindcss 로 동적 스타일링을 했는지 다른 라이브러리에서  봐둔게 있는게 그거를 대면 회의에서 설명해도될까?
[headlessUI](https://headlessui.com/)라는 컴포넌트 라이브러리에서 본 방식인데 tailwindcss로 스타일함에도 불구하고 컴포넌트의 동적 스타일을 구현해 놓은 부분이 있어서 이 부분을 한번 이야기 해보려고 해!

**간단하게 하자면** html의 [`data attribute`](https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors)로 컴포넌트의 상태에 따라 다른 UI가 보이도록, 그니까 동적으로 스타일링 할 수 있도록 구현했더라고! headlessUI 뿐만 아니라 [RadixUI](https://www.radix-ui.com/)도 마찬가지야.

그래서 살짝 내 생각의 결론을 말하자면,
> [headlessUI](https://headlessui.com/)나 [RadixUI](https://www.radix-ui.com/)같은 `unstyled` 컴포넌트 라이브러리를 참고하거나 가져다 사용한다.

인데 이 부분도 자세히 설명해줄게~


### 방법 4. 커스텀 클래스 정의
위에서도 말했지만 tailwindcss 문서에서도 커스텀 클래스를 만드는 것을 엄청 권장하고 있지는 않아보여.
> If you start using `@apply` for everything, you are basically just writing CSS again and throwing away all of the workflow and maintainability advantages Tailwind gives you
> https://tailwindcss.com/docs/reusing-styles#avoiding-premature-abstraction

라고 말하고 있거든!

**Reply by elbica** [reply]
다 맞는 소리야!! 특히 data attribute로 동적 스타일링 구현한게 엄청 궁금하네!!

`1번` 같은 경우는 prop 타입 체크가 돼서 일단 저렇게 사용하긴 했었는데 좋은 방식 같진 않아보여..🥲
- tailwind에서 커스텀 스타일을 남용하지 말라는 이유와 비슷

그리고 중복보다는 2번의 복잡한 가독성때문에 key로 접근하는 방식을 사용했었어!


`2번`도 예시가 맞는데 다른 레포들 보면 거의 다 삼항 연산자 같은 걸로 분기처리를 저렇게 했더라고 (이것도 가독성이 안좋아보임)

그런데 twin macro랑 data attribute는 사용성에 좀 차이가 있겠다!
twin은 자료도 있고 css in js 동적 스타일처럼 사용성도 기존보다 좋아보이는데
data attribute 방식은 익숙하지가 않아서..🙃 그래도 도전해보면 되게 재밌겠다
